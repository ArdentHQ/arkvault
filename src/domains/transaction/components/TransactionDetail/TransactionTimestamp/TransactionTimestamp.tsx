import { DateTime } from "@payvo/sdk-intl";
import React from "react";
import { useTranslation } from "react-i18next";

import { useTimeFormat } from "@/app/hooks/use-time-format";
import { TransactionDetail } from "@/domains/transaction/components/TransactionDetail/TransactionDetail";

interface TransactionTimestampProperties {
	timestamp?: DateTime;
}

export const TransactionTimestamp = ({ timestamp }: TransactionTimestampProperties) => {
	const { t } = useTranslation();
	const format = useTimeFormat();

	return (
		<TransactionDetail data-testid="TransactionTimestamp" label={t("TRANSACTION.TIMESTAMP")}>
			{timestamp?.format(format)}
		</TransactionDetail>
	);
};
