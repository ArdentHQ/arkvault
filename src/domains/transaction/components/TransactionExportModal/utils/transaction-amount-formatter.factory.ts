import { BigNumber } from "@ardenthq/sdk-helpers";
import { DTO, Helpers } from "@ardenthq/sdk-profiles";

const formatAmount = (amount: number, transaction: DTO.ExtendedConfirmedTransactionData) =>
	Helpers.Currency.format(amount, transaction.wallet().exchangeCurrency(), {
		withTicker: false,
	});

const multiPaymentAmount = (transaction: DTO.ExtendedConfirmedTransactionData) => {
	if (transaction.isReceived()) {
		let totalReceived = BigNumber.make(transaction.amount());

		for (const recipient of transaction.recipients()) {
			if (recipient.address === transaction.wallet().address()) {
				totalReceived.minus(recipient.amount);
			}
		}

		return formatAmount(totalReceived.toNumber(), transaction);
	}

	return formatAmount(transaction.amount(), transaction);
};

const transactionAmount = (transaction: DTO.ExtendedConfirmedTransactionData) => {
	if (transaction.isMultiPayment()) {
		return multiPaymentAmount(transaction);
	}

	return formatAmount(transaction.amount(), transaction);
};

const transactionTotal = (transaction: DTO.ExtendedConfirmedTransactionData) => {
	if (transaction.isMultiPayment()) {
		return BigNumber.make(multiPaymentAmount(transaction)).plus(transaction.fee());
	}

	return formatAmount(transaction.total(), transaction);
};

export const AmountFormatter = (transaction: DTO.ExtendedConfirmedTransactionData) => ({
	amount: () => transactionAmount(transaction),
	fee: () => formatAmount(transaction.fee(), transaction),
	total: () => transactionTotal(transaction),
});
