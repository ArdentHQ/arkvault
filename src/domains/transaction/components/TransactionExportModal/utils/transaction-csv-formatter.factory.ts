import { BigNumber } from "@ardenthq/sdk-helpers";
import { DTO } from "@ardenthq/sdk-profiles";
import { CURRENCIES } from "@ardenthq/sdk-intl";
import { buildTranslations } from "@/app/i18n/helpers";

const recipient = (transaction: DTO.ExtendedConfirmedTransactionData) => {
	const { COMMON } = buildTranslations();

	if (transaction.isMultiPayment()) {
		return COMMON.MULTIPLE;
	}

	if (transaction.isTransfer()) {
		return transaction.recipient();
	}

	if (transaction.isVote() || transaction.isUnvote()) {
		return `${COMMON.VOTE} ${COMMON.TRANSACTION}`;
	}

	return COMMON.OTHER;
};

const multiPaymentAmount = (transaction: DTO.ExtendedConfirmedTransactionData): number => {
	if (transaction.sender() !== transaction.wallet().address()) {
		let totalReceived = BigNumber.ZERO;

		for (const recipient of transaction.recipients()) {
			if (recipient.address === transaction.wallet().address()) {
				totalReceived = totalReceived.plus(recipient.amount);
			}
		}

		return totalReceived.toNumber();
	}

	return transaction.amount();
};

const transactionAmount = (transaction: DTO.ExtendedConfirmedTransactionData): number => {
	const amount = transaction.isMultiPayment() ? multiPaymentAmount(transaction) : transaction.amount();

	if (transaction.isSent()) {
		return BigNumber.make(amount).times(-1).toNumber();
	}

	return amount;
};

const transactionFee = (transaction: DTO.ExtendedConfirmedTransactionData): number => {
	if (transaction.isSent()) {
		return BigNumber.make(transaction.fee()).times(-1).toNumber();
	}

	return 0;
};

const transactionTotal = (transaction: DTO.ExtendedConfirmedTransactionData): number => {
	if (transaction.isSent()) {
		return BigNumber.make(transaction.total()).times(-1).toNumber();
	}

	return transactionAmount(transaction);
};

const converted = (value: number, rate: number) => BigNumber.make(value).times(rate).toNumber();

const truncate = (value: number, currency: string) => {
	const decimals = CURRENCIES[currency]?.decimals ?? 8;

	return Math.trunc(value * 10 ** decimals) / 10 ** decimals;
};

export const CsvFormatter = (transaction: DTO.ExtendedConfirmedTransactionData, timeFormat: string) => {
	const amount = transactionAmount(transaction);
	const fee = transactionFee(transaction);
	const total = transactionTotal(transaction);

	const currency = transaction.wallet().currency();
	const exchangeCurrency = transaction.wallet().exchangeCurrency();

	const rate =
		transaction.total() === 0
			? 0
			: truncate(
					BigNumber.make(transaction.convertedTotal()).divide(transaction.total()).toNumber(),
					exchangeCurrency,
			  );

	return {
		amount: () => truncate(amount, currency),
		convertedAmount: () => truncate(converted(amount, rate), exchangeCurrency),
		convertedFee: () => (fee === 0 ? 0 : truncate(converted(fee, rate), exchangeCurrency)),
		convertedTotal: () => truncate(converted(total, rate), exchangeCurrency),
		datetime: () => transaction.timestamp()?.format(`DD.MM.YYYY ${timeFormat}`),
		fee: () => fee,
		rate: () => rate,
		recipient: () => recipient(transaction),
		sender: () => transaction.sender(),
		timestamp: () => transaction.timestamp()?.toUNIX(),
		total: () => truncate(total, currency),
	};
};
