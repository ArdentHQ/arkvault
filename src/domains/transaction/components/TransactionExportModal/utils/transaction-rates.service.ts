import { Contracts } from "@ardenthq/sdk-profiles";
import { DateTime } from "@ardenthq/sdk-intl";
import { BigNumber } from "@ardenthq/sdk-helpers";
import { HttpClient } from "@/app/services/HttpClient";

export const TransactionRates = ({ wallet }: { wallet: Contracts.IReadWriteWallet }) => {
	const exchangeEndpoint = "https://min-api.cryptocompare.com/data/histoday";
	const client = new HttpClient(0);
	const baseCurrency = "USD";

	const rates = {
		[baseCurrency]: new Map(),
	};

	const sync = async ({ from = DateTime.make(), to }: { from?: DateTime; to: DateTime }) => {
		const response = await client.get(exchangeEndpoint, {
			fsym: wallet.network().ticker(),
			limit: from.diffInDays(to),
			toTs: to.toUNIX(),
			tsym: baseCurrency,
		});

		for (const price of response.json().Data) {
			const ymd = DateTime.fromUnix(price.time).format("YYYY-MM-DD");
			rates[baseCurrency].set(ymd, BigNumber.make(price.close).toString());
		}
	};

	const byTimestamp = (timestamp?: DateTime) =>
		rates[baseCurrency].get(DateTime.make(timestamp).format("YYYY-MM-DD"));

	return {
		rate: (timestamp?: DateTime) => byTimestamp(timestamp),
		sync,
	};
};
