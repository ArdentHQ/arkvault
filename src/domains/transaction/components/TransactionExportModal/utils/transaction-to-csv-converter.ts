import { DTO } from "@ardenthq/sdk-profiles";
import { CsvSettings } from "@/domains/transaction/components/TransactionExportModal";
import { buildTranslations } from "@/app/i18n/helpers";
import { AmountFormatter } from "./transaction-amount-formatter.factory";

const headers = (settings: CsvSettings) => {
	const { COMMON } = buildTranslations();

	return [
		...(settings.includeTransactionId ? [COMMON.ID] : []),
		...(settings.includeDate ? [COMMON.TIMESTAMP, COMMON.DATE] : []),
		...(settings.includeSenderRecipient ? [COMMON.SENDER, COMMON.RECIPIENT] : []),
		...(settings.includeCryptoAmount ? [COMMON.AMOUNT, COMMON.FEE, COMMON.TOTAL] : []),
	].join(settings.delimiter);
};

const transactionToCsv = (transaction: DTO.ExtendedConfirmedTransactionData, settings: CsvSettings) => {
	const dateTimeFormat = "DD.MM.YYYY h:mm A";
	const timestamp = transaction.timestamp()?.toUNIX();
	const datetime = transaction.timestamp()?.format(dateTimeFormat);
	const sender = transaction.sender();
	const recipient = transaction.recipient();
	const amounts = AmountFormatter(transaction);

	const columns = [
		...(settings.includeTransactionId ? [transaction.id()] : []),
		...(settings.includeDate ? [timestamp, datetime] : []),
		...(settings.includeSenderRecipient ? [sender, recipient] : []),
		...(settings.includeCryptoAmount ? [amounts.amount(), amounts.fee(), amounts.total()] : []),
	];

	return columns.join(settings.delimiter);
};

export const convertToCsv = (transactions: DTO.ExtendedConfirmedTransactionData[], settings: CsvSettings) => {
	const rows = transactions.map((transaction) => transactionToCsv(transaction, settings));

	if (!settings.includeHeaderRow) {
		return rows.join("\n");
	}

	return [...[headers(settings)], ...rows].join("\n");
};
