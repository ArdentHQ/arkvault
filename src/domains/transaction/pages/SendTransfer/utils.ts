export const getFeeType = (numberOfRecipients: number): "multiPayment" | "transfer" =>
	numberOfRecipients > 1 ? "multiPayment" : "transfer";
