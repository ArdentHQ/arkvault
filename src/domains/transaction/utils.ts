import { Services } from "@payvo/sdk";

export const isNoDeviceError = (error: any) => {
	if (!error) {
		return false;
	}

	if (String(error).includes("no device found")) {
		return true;
	}

	return false;
};

export const isRejectionError = (error: any) => String(error).includes("Condition of use not satisfied");

export const handleBroadcastError = ({ errors }: Services.BroadcastResponse) => {
	const allErrors = Object.values(errors);

	if (allErrors.length === 0) {
		return;
	}

	throw new Error(allErrors[0]);
};

export const withAbortPromise =
	(signal?: AbortSignal, callback?: () => void) =>
	<T>(promise: Promise<T>) =>
		new Promise<T>((resolve, reject) => {
			if (signal) {
				signal.addEventListener("abort", () => {
					callback?.();
					reject("ERR_ABORT");
				});
			}

			return promise.then(resolve).catch(reject);
		});
