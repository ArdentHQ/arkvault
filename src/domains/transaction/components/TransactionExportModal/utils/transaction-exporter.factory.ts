import { Contracts, DTO } from "@ardenthq/sdk-profiles";
import { Services } from "@ardenthq/sdk";
import { BigNumber } from "@ardenthq/sdk-helpers";
import { useState } from "react";
import { convertToCsv } from "./transaction-to-csv-converter";
import { CsvSettings } from "@/domains/transaction/components/TransactionExportModal";
import { assertString } from "@/utils/assertions";

interface TransactionExporterFetchProperties {
	type: "all" | "received" | "sent";
	dateRange?: Services.RangeCriteria;
	cursor?: number;
}

const filterTransactions = (transactions: DTO.ExtendedConfirmedTransactionData[]) =>
	transactions.filter((transaction) => {
		if (transaction.isTransfer()) {
			return transaction.sender() !== transaction.recipient();
		}

		if (transaction.isMultiPayment()) {
			let amount = BigNumber.make(transaction.amount());

			for (const recipient of transaction.recipients()) {
				if (transaction.sender() === recipient.address) {
					amount = amount.minus(recipient.amount);
				}
			}

			return !amount.isZero();
		}

		return false;
	});

export const TransactionExporter = ({
	profile,
	wallet,
	limit = 100,
}: {
	profile: Contracts.IProfile;
	wallet: Contracts.IReadWriteWallet;
	limit?: number;
}) => {
	const exchangeCurrency = profile.settings().get<string>(Contracts.ProfileSetting.ExchangeCurrency);
	const timeFormat = profile.settings().get<string>(Contracts.ProfileSetting.TimeFormat);

	let count = 0;

	assertString(exchangeCurrency);
	assertString(timeFormat);

	let requestedSyncAbort = false;

	let transactions: DTO.ExtendedConfirmedTransactionData[] = [];

	const sync = async ({
		type = "all",
		dateRange,
		cursor = 1,
	}: TransactionExporterFetchProperties): Promise<number | undefined> => {
		// Clear cache.
		if (cursor === 1) {
			transactions = [];
			requestedSyncAbort = false;
		}

		if (requestedSyncAbort) {
			return;
		}

		const page = await wallet.transactionIndex()[type]({ cursor, limit, timestamp: dateRange });

		transactions.push(...page.items());
		cursor = cursor + 1;

		count = transactions.length;

		// Last page.
		// TODO: Not relying on totalCount because it is an estimate
		//        and is not giving accurate pagination info. Address this issue after initial implementation.
		if (page.items().length < limit) {
			if (type === "received") {
				transactions = filterTransactions(transactions);
			}

			return transactions.length;
		}

		return sync({ cursor, dateRange, type });
	};

	return {
		transactions: () => ({
			abortSync: () => (requestedSyncAbort = true),
			count: () => count,
			items: () => transactions,
			sync,
			toCsv: (settings: CsvSettings) => convertToCsv(transactions, settings, exchangeCurrency, timeFormat),
		}),
	};
};
