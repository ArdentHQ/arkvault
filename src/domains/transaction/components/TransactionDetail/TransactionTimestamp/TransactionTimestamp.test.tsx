import { DateTime } from "@payvo/sdk-intl";
import React from "react";

import { TransactionTimestamp } from "./TransactionTimestamp";
import { render } from "@/utils/testing-library";

const datetime = DateTime.fromUnix(1_596_213_281);

describe("TransactionTimestamp", () => {
	it("should render", () => {
		const { container } = render(<TransactionTimestamp timestamp={datetime} />);

		expect(container).toHaveTextContent("31.07.2020 4:34 PM");
		expect(container).toMatchSnapshot();
	});
});
