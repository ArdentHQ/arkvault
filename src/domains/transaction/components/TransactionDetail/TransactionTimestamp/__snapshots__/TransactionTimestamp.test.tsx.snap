// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionTimestamp should render 1`] = `
<div>
  <div
    class="no-ligatures css-19rivvh"
    data-testid="TransactionTimestamp"
  >
    <div
      class="w-40 flex-1 space-y-2 whitespace-nowrap"
      data-testid="TransactionDetail--desktop"
    >
      <div
        class="no-ligatures text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
      >
        Timestamp
      </div>
      <div
        class="flex items-center font-semibold"
      >
        31.07.2020 4:34 PM
      </div>
    </div>
  </div>
</div>
`;
